python-lib/osisoft_client.py

```diff
@@ -43,6 +43,144 @@ def get_auth(self, auth_type, username, password):
         else:
             return None
 
+    def get_all_rows_from_webid(self, webid, data_type, start_date=None, end_date=None,
+                                interval=None, sync_time=None, boundary_type=None, selected_fields=None,
+                                can_raise=True, endpoint_type="event_frames", search_full_hierarchy=None, max_count=None):
+        # Split the time range until no more HTTP 400
+        not_happy = True
+        previous_item_timestamp = False
+        while not_happy:
+            logger.info("Attempting download from {} to {}".format(start_date, end_date))
+            ret = self.get_row_from_webid(webid, data_type, start_date=start_date, end_date=end_date,
+                                          interval=interval, sync_time=sync_time, boundary_type=boundary_type, selected_fields=selected_fields,
+                                          can_raise=can_raise, endpoint_type=endpoint_type, search_full_hierarchy=search_full_hierarchy, max_count=max_count)
+            """
+            ret = self.get_row_from_item(item, data_type, start_date=start_date, end_date=end_date, interval=interval,
+                          sync_time=sync_time, boundary_type=boundary_type, can_raise=can_raise, object_id=None,
+                          search_full_hierarchy=search_full_hierarchy, max_count=max_count):
+            """
+            counter = 0
+            try:
+                first_item = next(ret)
+                if not previous_item_timestamp or previous_item_timestamp != first_item.get("Timestamp"):
+                    logger.info("Yielding first item")
+                    yield first_item
+                counter += 1
+                for first_item in ret:
+                    yield first_item
+                    counter += 1
+            except Exception as err:
+                if "Error 400" in "{}".format(err):
```

Comment:
Bug:
This error check is a bit general - we get an "Error 400" for other things, e.g. when the interval param is wrong (`s` or `1sec`). Then it goes into an infinite loop and hits the max recursion depth of python:
![image](https://github.com/user-attachments/assets/6cfeb836-fa92-457b-8d53-248fccaaf3b4)

This is a rather dangerous and obscures the actual error from the user.

I think it would be best to either:
1. make this check pretty specific - e.g. check for the string `Parameter 'timeRange / intervals' is greater than the maximum allowed`
2. find a way to do this that doesn't involve recursion - e.g. work out the time ranges needed to be under the 150,000 limit, then batch up the requests evenly across the time period and do it in a loop.

These strategies might only possible if we are only concerned with Interpolation here. But I think that must be case. It's clear we expect to get back a 400 status when there are too many records, which implies a problem in the request, not in the recorded data set being too large. I imagine if the recorded data exceeds 150,000 records it will be limited by the max_count and you already have something else to handle that. So I think it is only interpolation that has this problem this bit of code addresses.

Reply:
👍 Fixed in [3a66cfc](https://github.com/dataiku/dss-plugin-osisoft/pull/43/commits/3a66cfcb4dcca84d46346e3a1fd584faa113dbb4)

Reply:
Nice, handles that error much better now